app/javascript/mastodon/components/column_header.tsx

```diff
@@ -0,0 +1,309 @@
+import PropTypes from 'prop-types';
+import { PureComponent, useCallback } from 'react';
+
+import { FormattedMessage, injectIntl, defineMessages } from 'react-intl';
+import type { IntlShape } from 'react-intl';
+
+import classNames from 'classnames';
+import type { RouteComponentProps } from 'react-router-dom';
+import { withRouter } from 'react-router-dom';
+
+import { ReactComponent as AddIcon } from '@material-symbols/svg-600/outlined/add.svg';
+import { ReactComponent as ArrowBackIcon } from '@material-symbols/svg-600/outlined/arrow_back.svg';
+import { ReactComponent as ChevronLeftIcon } from '@material-symbols/svg-600/outlined/chevron_left.svg';
+import { ReactComponent as ChevronRightIcon } from '@material-symbols/svg-600/outlined/chevron_right.svg';
+import { ReactComponent as CloseIcon } from '@material-symbols/svg-600/outlined/close.svg';
+import { ReactComponent as TuneIcon } from '@material-symbols/svg-600/outlined/tune.svg';
+
+import type { IconProp } from 'mastodon/components/icon';
+import { Icon } from 'mastodon/components/icon';
+import {
+  ButtonInTabsBar,
+  useColumnsContext,
+} from 'mastodon/features/ui/util/columns_context';
+import type { ContextWithIdentity } from 'mastodon/utils/identity';
+
+import { useAppHistory } from './router';
+
+const messages = defineMessages({
+  show: { id: 'column_header.show_settings', defaultMessage: 'Show settings' },
+  hide: { id: 'column_header.hide_settings', defaultMessage: 'Hide settings' },
+  moveLeft: {
+    id: 'column_header.moveLeft_settings',
+    defaultMessage: 'Move column to the left',
+  },
+  moveRight: {
+    id: 'column_header.moveRight_settings',
+    defaultMessage: 'Move column to the right',
+  },
+});
+
+interface BackButtonProps {
+  pinned?: boolean;
+  show?: boolean;
+}
+
+const BackButton = ({ pinned, show }: BackButtonProps) => {
+  const history = useAppHistory();
+  const { multiColumn } = useColumnsContext();
+
+  const handleBackClick = useCallback(() => {
+    if (history.location.state?.fromMastodon) {
+      history.goBack();
+    } else {
+      history.push('/');
+    }
+  }, [history]);
+
+  const showButton =
+    // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
+    history &&
```

Comment:
Maybe we should change `useAppHistory` to return `History | undefined`, if it may not be always defined?

Reply:
The `react-router` types indicate this should always be defined. Is it known for sure that this can be undefined?

Reply:
This is complex, as some components are mounted in the admin panel (for example in the moderation interface), which is not wrapped into a `<Router>` component because those pages are Rails views. Those components currently use `withOptionalRouter` to indicate that those props might not be defined if rendered outside of the router.

I guess we can have a `useAppOptionalHistory()`, which will be used in components where `withOptionalRouter` is currently used, and will return `History | undefined` so it is handled properly.

This is not a good long-term solution, but will be equivalent to what we currently have?

Reply:
Yeah the `useAppOptionalHistory()` idea is what I'd gravitate to personally.